### ISSUE TITLE
local storage provisioner reads wrong capacity

### ISSUE THREAD

**Issue by jsafrane** [kubernetes-incubator/external-storage#370]
Let's pretend we have mount propagation in place and provisioner can see mounts that were mounted while the provisioner was running.

There is a race between admin mounting a volume and provisioner creating a PV for it:

1. admin creates /mnt/disks/new-volume1
2. admin mounts /dev/sdc to /mnt/disks/new-volume1

Provisioner will create a PV with wrong capacity when it scans /mnt/disks after step 1 and before 2. Should the provisioner check that new-volume1 is a mount point before creating the PV? It won't be possible to create a PV for the root partition then. Should admin bind-mount the root partition when creating a PV for the root partition is really wanted?

@wongma7 @ddysher @msau42 @ianchakeres, WDYT?

**Comment by ianchakeres** [comment]
Here are three possible solutions:

#1 - Admin use permissions

Have the provisioner ignore directories with specific restricted permissions and have the administrator set these permissions on any directory (/mnt/disks/new-volume1) that should be ignored by the provisioner.

When the mount operation is performed, the now mounted location's permissions (/dev/sdc) will take over the directory permissions.

If we wanted to further restrict the directories permissions that are considered ignored, then we could also use the CAP_LINUX_IMMUTABLE on the mount point directory (/mnt/disks/new-volume1). This parameter could be checked by the provisioner, and the provisioner could avoid creating local PV for any directory with this flag set. 

#2 - Admin uses lockfile 

We could have a lock file that the administrator can set/clear to avoid the provisioner from scanning a directory for potential PV or a specific lock file to exclude a particular directory (e.g. new-volume1.lock).

#3 - Admin stops provisioner

The administrator could be expected to stop the provisioner during these types of operations.

**Comment by wongma7** [comment]
IMO yes and yes, it would be good to mandate that every directory in /mnt/disks is a mountpoint. That requirement is clear, it doesn't require the admin to understand any new mechanism, and forcing the admin to do a bind mount to explicitly express that "this dir is ready to have a pv created for it" is less work than any of the other solutions.

**Comment by msau42** [comment]
Unfortunately as part of our user requirements, we also need to support directories, and not just mount points.  Another possibility is to only allow specific types of volumes per directory and add it to the configuration:

storageClass = "fast", hostDir = "/mnt/ssds", type = "mounts"
storageClass = "shared", hostDir = "/something/in/root", type = "directories"
storageClass = "ssds", hostDir = "/dev/disks/somewhere", type = "block"

Then the provisioner discovery method will be different for every directory depending on the type.

**Comment by ddysher** [comment]
How do we handle config update if we add the new 'type' parameter? e.g. admin changed his mind and want to make /mnt/ssds to be shared directory.  The directory already have mount points and have corresponding PVs created in kubernetes, should we maintain these 'legacy' PVs?

**Comment by jsafrane** [comment]
Anything that we choose must be 1) simple and 2) must not lead to wrong PVs being created.

@ianchakeres, all your proposal might lead to creation of wrong PV if the admin is careless (and they are careless). If they forgot to create .lock file or forgot that their umask is wrong or forgot to stop the provisioner the resulting PV would have wrong capacity, something may bind to it and put some data there. There is no easy way how to fix it later.

What's wrong with bind-mounts? As @wongma7 pointed out, by bind-mounting a subdirectory admin says "this directory is ready to become a PV". You can bind-mount the same directory to the same place,
 there is nothing wrong about it:

```
mount --bind /mnt/disks/volume1 /mnt/disks/volume1
```

The only downside of mount points is parsing of `/proc/self/mountinfo`, it can be really big on some systems. Especially on bare metal.

**Comment by msau42** [comment]
I'm fine with requiring filesystem volumes to be bind mounted, and block volumes to be symlinked.  For some reason I thought there was some concern about the extra step of bind mounting, but I misread.

I'm ok with the detection potentially taking a long time.  Discovery of local volumes doesn't need to be that performant.  We can at least cache the mounts for each iteration so that we don't need to call it for every single volume.

Changing this requirement is going to be a breaking change.  When older PVs are deleted, they're not going to get rediscovered.  Considering this feature is still alpha, I am also fine with that.  We just need to document it clearly in the changelog, and probably increment a major version.

**Comment by msau42** [comment]
Matt has graciously offered to look into this. Thanks!

/assign @wongma7

**Comment by ianchakeres** [comment]
I've got another idea...

Have the local-volume provisioner make two passes over the discovery directory before creating a PV for a new directory. The first pass would detect that there is a new directory (e.g /mnt/disks/vol1). The provisioner waits some period of time long enough for us to reasonably expect that an admin has completed configuration of the new directory (e.g. creating mount point and mounting volume to it). The second pass detects that the new directory still exists, and it creates the related PV.

@jsafrane @msau42  @wongma7 - what do you think of this approach?

**Comment by ddysher** [comment]
How long is enough for admin to complete the mount operation?  I feel it only decreases the changes of race condition, the problem is inherently the same.  This also requires admin to be aware of how provisioner works internally.

**Comment by ianchakeres** [comment]
Another issue related to race condition when creating mountpoint for discovery, https://github.com/kubernetes-incubator/external-storage/issues/482.

**Comment by ianchakeres** [comment]
@msau42 - I think the agreed upon the plan to handle the race condition was to check whether a discovered path is an actual mountpoint before creating a related local PV, based upon our offline conversation. 

Do we have agreement on this solution? 

@msau42 @jsafrane @ddysher @dhirajh @wongma7 WDYT?

If yes, I'll get started on a PR.

**Comment by msau42** [comment]
Yes, but before you can do that, you'll also have to enable and configure mount propagation in the daemonset hostpath volumes.

**Comment by jsafrane** [comment]
/close

Fixed by #499
